Mathlib/Analysis/Normed/Group/Ultra.lean

```diff
@@ -0,0 +1,175 @@
+
+/-- All triangles are isosceles in an ultrametric normed group. -/
+@[to_additive "All triangles are isosceles in an ultrametric normed additive group."]
+lemma nnnorm_mul_eq_max_of_nnnorm_ne_nnnorm
+    {x y : S} (h : ‖x‖₊ ≠ ‖y‖₊) : ‖x * y‖₊ = max ‖x‖₊ ‖y‖₊ := by
+  simpa only [← NNReal.coe_inj, NNReal.coe_max] using
+    norm_mul_eq_max_of_norm_ne_norm (NNReal.coe_injective.ne h)
+
+@[to_additive]
+lemma nnnorm_eq_of_mul_nnnorm_lt_max {x y : S} (h : ‖x * y‖₊ < max ‖x‖₊ ‖y‖₊) :
+    ‖x‖₊ = ‖y‖₊ :=
+  not_ne_iff.mp (h.ne ∘ nnnorm_mul_eq_max_of_nnnorm_ne_nnnorm)
+
+/-- All triangles are isosceles in an ultrametric normed group. -/
+@[to_additive "All triangles are isosceles in an ultrametric normed additive group."]
+lemma norm_div_eq_max_of_norm_div_ne_norm_div (x y z : S) (h : ‖x / y‖ ≠ ‖y / z‖) :
+    ‖x / z‖ = max ‖x / y‖ ‖y / z‖ := by
+  simpa only [div_mul_div_cancel] using norm_mul_eq_max_of_norm_ne_norm h
+
+/-- All triangles are isosceles in an ultrametric normed group. -/
+@[to_additive "All triangles are isosceles in an ultrametric normed additive group."]
+lemma nnnorm_div_eq_max_of_nnnorm_div_ne_nnnorm_div (x y z : S) (h : ‖x / y‖₊ ≠ ‖y / z‖₊) :
+    ‖x / z‖₊ = max ‖x / y‖₊ ‖y / z‖₊ := by
+  simpa only [← NNReal.coe_inj, NNReal.coe_max] using
+    norm_div_eq_max_of_norm_div_ne_norm_div _ _ _ (NNReal.coe_injective.ne h)
+
+@[to_additive]
+lemma nnnorm_pow_le (x : S) (n : ℕ) :
+    ‖x ^ n‖₊ ≤ ‖x‖₊ := by
+  induction n with
+  | zero => simp
+  | succ n hn => simpa [pow_add, hn] using nnnorm_mul_le_max (x ^ n) x
+
+@[to_additive]
+lemma norm_pow_le (x : S) (n : ℕ) :
+    ‖x ^ n‖ ≤ ‖x‖ :=
+  nnnorm_pow_le x n
+
+@[to_additive]
+lemma nnnorm_zpow_le (x : S) (z : ℤ) :
+    ‖x ^ z‖₊ ≤ ‖x‖₊ := by
+  cases z <;>
+  simpa using nnnorm_pow_le _ _
+
+@[to_additive]
+lemma norm_zpow_le (x : S) (z : ℤ) :
+    ‖x ^ z‖ ≤ ‖x‖ :=
+  nnnorm_zpow_le x z
+
+end Group
+
+section CommGroup
+
+variable {M ι : Type*} [SeminormedCommGroup M] [IsUltrametricDist M]
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product.
+This version is phrased using `Finset.sup'` and `Finset.Nonempty` due to `Finset.sup`
+operating over an `OrderBot`, which `ℝ` is not.
+-/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum.
+This version is phrased using `Finset.sup'` and `Finset.Nonempty` due to `Finset.sup`
+operating over an `OrderBot`, which `ℝ` is not. "]
+lemma _root_.Finset.Nonempty.norm_prod_le_sup'_norm {s : Finset ι} (hs : s.Nonempty) (f : ι → M) :
+    ‖∏ i ∈ s, f i‖ ≤ s.sup' hs (‖f ·‖) := by
+  simp only [Finset.le_sup'_iff]
+  refine Finset.Nonempty.cons_induction ?_ ?_ hs
+  · simp
+  rintro a s ha _ ⟨b, hb, IH⟩
+  simp only [Finset.prod_cons, Finset.mem_cons, exists_eq_or_imp]
+  refine (le_total ‖∏ i ∈ s, f i‖ ‖f a‖).imp ?_ ?_ <;> intro h
+  · exact (norm_mul_le_max _ _).trans (by simp [h])
+  · exact ⟨b, hb, (norm_mul_le_max _ _).trans (by simpa [h] using IH)⟩
```

Comment:
You can write this in a more "lean4" syntax as follows:
```lean
lemma _root_.Finset.Nonempty.norm_prod_le_sup'_norm {s : Finset ι} (hs : s.Nonempty) (f : ι → M) :
    ‖∏ i ∈ s, f i‖ ≤ s.sup' hs (‖f ·‖) := by
  simp only [Finset.le_sup'_iff]
  induction hs using Finset.Nonempty.cons_induction with
  | singleton j => simp only [Finset.mem_singleton, Finset.prod_singleton, exists_eq_left, le_refl]
  | cons j t hj _ IH =>
      simp only [Finset.prod_cons, Finset.mem_cons, exists_eq_or_imp]
      rcases le_total ‖∏ i ∈ t, f i‖ ‖f j‖ with h | h
      · exact .inl <| (norm_mul_le_max _ _).trans (max_eq_left h).le
      · rcases IH with ⟨b, hb, IH⟩
        exact .inr ⟨b, hb, (norm_mul_le_max _ _).trans <| (max_eq_right h).le.trans IH⟩
```
(One line longer than the old version, since there's no equivalent of `rintro` in the new induction syntax... 🏌️ )